### PR TITLE
Principles for testing CumulusCI.

### PR DESCRIPTION
**Pull request by prescod**
I think of this as an extension of our contributor docs, but I'll get some feedback before I link it into the main docset.

### CODE REVIEW

File: docs/testing-principles.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
+# Principles for Testing the Cumulus Framework and Snowfakery
+
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:2]: Also not a bad idea to have a sentence or two here.

File: docs/testing-principles.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+## Goals
+
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:4]:
```suggestion

Here are the goals for testing:

```

File: docs/testing-principles.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+1. tests should execute code paths as similar to end-user code as possible (fidelity)
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:5]:
```suggestion
* Tests should execute code paths as similar to end-user code as possible (fidelity)
```

File: docs/testing-principles.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+2. tests should run as fast as possible (velocity)
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:6]:
```suggestion
* Tests should run as fast as possible (velocity)
```

File: docs/testing-principles.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+3. tests should run properly even when third-party services are down or unavailable (availability)
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:7]:
```suggestion
* Tests should run properly even when third-party services are down or unavailable (availability)
```

File: docs/testing-principles.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+Thus, mocking/faking should be used sparingly (as they work against
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:9]:
```suggestion
Mocking and faking work against the goal of fidelity, so only 
```

File: docs/testing-principles.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+the goal of fidelity) and should only to deal with challenges
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:10]:
```suggestion
use them when necessary to deal with challenges of velocity and availability.
```

File: docs/testing-principles.md [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+of velocity and availability.
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:11]:
```suggestion

```

File: docs/testing-principles.md [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+## VCR for the easy cases
+
+For simple cases, a transformation to use VCR allows all three goals to
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:15]:
```suggestion
For simple cases, a transformation to use VCR helps achieve
```

File: docs/testing-principles.md [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+be simultaneously achieved. Now that we have VCR, we can take a test like this:
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:16]:
```suggestion
all three goals. Now that we have VCR, we can take a test like this:
```

File: docs/testing-principles.md [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    @pytest.mark.vcr()
+    def test_license_preflight(self, create_task):
+        task = create_task(GetAvailableLicenses, {})
+        task()
+        expected = set(["SFDC", "CLOUD_INTEGRATION_USER"])
+        actual = set(task.return_values)
+        assert not expected.difference(actual), task.return_values
+```
+
+This meets all three of our core goals, and also will:
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:50]:
```suggestion
This meets all three of our core goals, and also does the following:
```

File: docs/testing-principles.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+a) run against real orgs, when requested to
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:52]:
```suggestion
* Run against real orgs, when requested to
```

File: docs/testing-principles.md [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+b) run against pretend orgs, otherwise
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:53]:
```suggestion
* Run against pretend orgs, otherwise
```

File: docs/testing-principles.md [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+c) document the real-life behaviour of the task
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:54]:
```suggestion
* Document the real-life behavior of the task
```

File: docs/testing-principles.md [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+d) survive any reorganization of the internals of the task
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:55]:
```suggestion
* Survive any reorganization of the internals of the task
```

File: docs/testing-principles.md [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+
+The ability to run against real orgs is an important
+confidence-building measure when working with an environment
+that constantly changes like SAAS services.
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:59]:
```suggestion
that constantly changes, such as SAAS services.
```

File: docs/testing-principles.md [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+Using our testing infrastructure, you do so like this:
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:61]:
```suggestion
Using our testing infrastructure, you run against a real org like this:
```

File: docs/testing-principles.md [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+```s
+$ pytest -k test_license_preflight --org qa
+```
+
+### How it Works
+
+What happens under the cover is that the `create_task` fixture knows whether
+the test suite is being run against a real org or not. If so, it passes
+an `sf_org` object connected to a real Salesforce org. If not, it passes
+a fake `sf_org` which will emulate a `qa_org`.
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:72]:
```suggestion
a fake `sf_org` which emulates a `qa_org`.
```

File: docs/testing-principles.md [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+
+In a perfect world, we would achieve a very high level of test coverage
+when every fake/mock is turned "off" and all code is running against
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:75]:
```suggestion
when every fake/mock is turned _off_ and all code is running against
```

File: docs/testing-principles.md [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+real systems, just as we have high coverage when they are ruened on.
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:76]:
```suggestion
real systems, just as we have high coverage when they are turned on.
```

File: docs/testing-principles.md [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+### Component Tests
+
+One of the goals is to test our system as real users use it, which is why
+we use high-level APIs like `task()` instead of low level ones like
+`task._run_task()`
+
+That said, a system as complicated as Cumulus Framework cannot be tested only
+from the _end_ user's perspective. Cumulus Framework consists of many components.
+
+The "users" of these components are programmers on our team, and they should
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:93]:
```suggestion
The users of these components are programmers on our team, and they should
```

File: docs/testing-principles.md [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+interchangably, depending on the current step in the SDLC.
+
+### Component Tests
+
+One of the goals is to test our system as real users use it, which is why
+we use high-level APIs like `task()` instead of low level ones like
+`task._run_task()`
+
+That said, a system as complicated as Cumulus Framework cannot be tested only
+from the _end_ user's perspective. Cumulus Framework consists of many components.
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:91]:
```suggestion
from the end user's perspective. Cumulus Framework consists of many components.
```

File: docs/testing-principles.md [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+The "users" of these components are programmers on our team, and they should
+be well-tested, just as if they were documented external APIs.
+
+For example, the YAML merge algorithm is a component. The XML manipulator is
+a component. The package XML library SHOULD BE a component, etc. These should
+have test cases that ensure that their interfaces do not accidentally change.
+
+## The Hard Cases: The Awkward Squad
+
+In the lingo of functional programmers, our tests doesn't deal with the
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:102]:
```suggestion
In the lingo of functional programmers, our tests don't deal with the
```

File: docs/testing-principles.md [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+"awkward squad" of issues other tests need to deal with:
+
+-   a variety of org shapes
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:105]:
```suggestion
-   A variety of org shapes
```

File: docs/testing-principles.md [hunk 25]
```diff
@@ -0,0 +1,200 @@
...
+-   error/exception conditions
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:106]:
```suggestion
-   Error/exception conditions
```

File: docs/testing-principles.md [hunk 26]
```diff
@@ -0,0 +1,200 @@
...
+-   mutable state (in the org)
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:107]:
```suggestion
-   Mutable state (in the org)
```

File: docs/testing-principles.md [hunk 27]
```diff
@@ -0,0 +1,200 @@
...
+-   delays, timestamps and other temporal aspects
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:108]:
```suggestion
-   Delays, timestamps and other temporal aspects
```

File: docs/testing-principles.md [hunk 28]
```diff
@@ -0,0 +1,200 @@
...
+orgs for tests:
+
+```
+@pytest.mark.vcr()
+@pytest.mark.org_shape("person_accounts", "config_person_accounts")
+def test_some_person_account_feature(self):
+    ...
+```
+
+This will run the `config_person_accounts` flow against the `person_accounts`
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:125]:
```suggestion
This runs the `config_person_accounts` flow against the `person_accounts`
```

File: docs/testing-principles.md [hunk 29]
```diff
@@ -0,0 +1,200 @@
...
+org shape and run the test in the context of that org. At least that's
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:126]:
```suggestion
org shape and runs the test in the context of that org. At least that's
```

File: docs/testing-principles.md [hunk 30]
```diff
@@ -0,0 +1,200 @@
...
+the goal. This infrastructure has not been tested or used much yet (e.g. for
+actual person accounts testing).
+
+### Simulating Networking Errors with VCR
+
+It is possible to simulate networking errors by hand-editing VCR files.
+This pattern can be observed [here](https://github.com/SFDO-Tooling/CumulusCI/blob/35b07ebfa80170f6e0e9d8cabd3da75bf41d5d59/cumulusci/utils/salesforce/tests/test_count_sobjects.py#L23) and [here](https://github.com/SFDO-Tooling/CumulusCI/blob/e3ba7e8b542755b6395dbc895daee3d1ea3d11dc/cumulusci/utils/salesforce/tests/cassettes/ManualEdit_TestCountSObjects.test_count_sobjects__network_errors.yaml#L16).
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:133]:
```suggestion
This pattern can be observed in [test_count_sobjects.py](https://github.com/SFDO-Tooling/CumulusCI/blob/35b07ebfa80170f6e0e9d8cabd3da75bf41d5d59/cumulusci/utils/salesforce/tests/test_count_sobjects.py#L23) and [ManualEdit_TestCountSObjects.test_count_sobjects__network_errors.yaml](https://github.com/SFDO-Tooling/CumulusCI/blob/e3ba7e8b542755b6395dbc895daee3d1ea3d11dc/cumulusci/utils/salesforce/tests/cassettes/ManualEdit_TestCountSObjects.test_count_sobjects__network_errors.yaml#L16).
```

Comment (pconrad-sfdo) [reply docs/testing-principles.md:133]: Not required, just nicer if people know where they're going. If there's a shorter way of expressing those links, that's fine too.

File: docs/testing-principles.md [hunk 31]
```diff
@@ -0,0 +1,200 @@
...
+### Service Connectors
+
+Third party services (including Salesforce) should always be wrapped in Service Connector
+abstractions.
+
+These abstractions can be tested with Requests and VCR.
+
+The overall business logic can be tested by injecting fakes for Service Connectors.
+
+A client library like simple_salesforce or github3 is not necessarily a good service
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:144]:
```suggestion
A client library like `simple_salesforce` or `github3` is not necessarily a good service
```

File: docs/testing-principles.md [hunk 32]
```diff
@@ -0,0 +1,200 @@
...
+connector abstraction, because they work at a protocol level and not a business logic
+level. Two examples, from simple_salesforce:
+
+1. We make heavy use of simple_salesforce.restful(), which tunnels MANY different types
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:148]:
```suggestion
* We make heavy use of `simple_salesforce.restful()`, which tunnels MANY different types
```

File: docs/testing-principles.md [hunk 33]
```diff
@@ -0,0 +1,200 @@
...
+   of requests through a single endpoint and is therefore hard to fake.
+
+2. We make heavy use of simple_salesforce.query() which has the same problem.
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:151]:
```suggestion
* We make heavy use of `simple_salesforce.query()`, which has the same problem.
```

File: docs/testing-principles.md [hunk 34]
```diff
@@ -0,0 +1,200 @@
...
+
+Github3 is better, but arguably if we had started using it by wrapping it we'd be in a
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:153]:
```suggestion
`Github3` is better, but arguably if we had started using it by wrapping it we'd be in a
```

File: docs/testing-principles.md [hunk 35]
```diff
@@ -0,0 +1,200 @@
...
+better position to add other VCS regardless, so maybe we should just always wrap
+client libraries by default. The case of Github3 is argably, but simple_salesforce
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:155]:
```suggestion
client libraries by default. The case of `Github3` is arguably, but `simple_salesforce`
```

Comment (pconrad-sfdo) [reply docs/testing-principles.md:155]: I'm not sure what this means. Above: `Github3` is better, maybe we should have wrapped it. Here: Github3 is ... arguably something? I don't know the difference you're calling out between these two libraries.

File: docs/testing-principles.md [hunk 36]
```diff
@@ -0,0 +1,200 @@
...
+is not: the fact that we do all sorts of protocol-level things (passing URLs,
+fiddling headers) is a clear giveawaay that we should have wrapped it.
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:157]: Should have wrapped which?

File: docs/testing-principles.md (hunk 34 above)
Comment (pconrad-sfdo) [reply docs/testing-principles.md:153]: This whole paragraph seems like an explanation of implementation thought process. Do we need it? If so, it's a little hard to follow.

File: docs/testing-principles.md [hunk 37]
```diff
@@ -0,0 +1,200 @@
...
+
+#### Wrap Universal Interfaces
+
+An abstraction like Salesforce.query() should be used for situations such as the
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:161]:
```suggestion
An abstraction like `Salesforce.query()` should be used for situations such as the
```

File: docs/testing-principles.md [hunk 38]
```diff
@@ -0,0 +1,200 @@
...
+end-user passing queries to us to execute. We, internally, should always wrap
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:162]: I don't know the audience, but in case it's external, do we need to tell them what we wish we were doing internally?

File: docs/testing-principles.md [hunk 39]
```diff
@@ -0,0 +1,200 @@
...
+a large component which calls `sf_org.query()` six (or six hundred!) times. You'll
+need to use string matching (which is intrinsically fragile) to figure out which
+return values to return for which call.
+
+Between the benefits of simpler tests, cleaner code and easier replacement of abstractions,
+ServiceConnectors will pay for themselves quickly.
+
+## Summarized Do's and Don'ts
+
+1. Use VCR when possible. Edit the VCR files if necessary to emulate networking failures.
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:187]:
```suggestion
* Use VCR when possible. Edit the VCR files if necessary to emulate networking failures.
```

File: docs/testing-principles.md [hunk 40]
```diff
@@ -0,0 +1,200 @@
...
+
+2. Mocking should only be used for replacing ServiceConnectors, not for
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:189]:
```suggestion
* Use mocking only for replacing ServiceConnectors, not for
```

File: docs/testing-principles.md [hunk 41]
```diff
@@ -0,0 +1,200 @@
...
+   replacing business logic code.
+
+3. URLs, header manipulations, etc. should only occur in ServiceConnectors
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:192]:
```suggestion
* URLs, header manipulations, etc. should only occur in ServiceConnectors
```

File: docs/testing-principles.md [hunk 42]
```diff
@@ -0,0 +1,200 @@
...
+   and never business logic.
+
+4. Tests should test user-visible logic OR component boundaries OR performance
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:195]:
```suggestion
* Tests should test user-visible logic OR component boundaries OR performance
```

File: docs/testing-principles.md [hunk 43]
```diff
@@ -0,0 +1,200 @@
...
+   logic (e.g. caching) and never the control flow within algorithms.
+
+5. Use "Universal Interfaces" like query languages and REST protocols only
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:198]:
```suggestion
* Use "Universal Interfaces" like query languages and REST protocols only
```

File: docs/testing-principles.md [hunk 44]
```diff
@@ -0,0 +1,200 @@
...
+   a) in ServiceConnectors or b) when the Universal Interface has been exposed
+   to the end-user themselves, and therefore cannot be wrapped in business logic.
```
Comment (pconrad-sfdo) [inline docs/testing-principles.md:200]:
```suggestion
   to the end-user and therefore can't be wrapped in business logic.
```

File: docs/testing-principles.md [hunk 45]
```diff
@@ -0,0 +1,200 @@
...
+This pattern can be observed [here](https://github.com/SFDO-Tooling/CumulusCI/blob/35b07ebfa80170f6e0e9d8cabd3da75bf41d5d59/cumulusci/utils/salesforce/tests/test_count_sobjects.py#L23) and [here](https://github.com/SFDO-Tooling/CumulusCI/blob/e3ba7e8b542755b6395dbc895daee3d1ea3d11dc/cumulusci/utils/salesforce/tests/cassettes/ManualEdit_TestCountSObjects.test_count_sobjects__network_errors.yaml#L16).
+
+### Service Connectors
+
+Third party services (including Salesforce) should always be wrapped in Service Connector
+abstractions.
+
+These abstractions can be tested with Requests and VCR.
+
+The overall business logic can be tested by injecting fakes for Service Connectors.
```
Comment (davidmreed) [inline docs/testing-principles.md:142]: This principle seems to me to be the highest level in this doc (subject to above notes). The decision tree looks like it runs:

- Are you using service connector model? 
  - If yes, test business logic with service connector fakes, no network access. Test service connector with VCR.
  - If no:
    - Prefer to test using VCR for anything that can be tested on an org shape that's reasonable to produce via automation
    - Test edge cases with mocking.

File: docs/testing-principles.md [hunk 46]
```diff
@@ -0,0 +1,200 @@
...
+```
+
+This will run the `config_person_accounts` flow against the `person_accounts`
+org shape and run the test in the context of that org. At least that's
+the goal. This infrastructure has not been tested or used much yet (e.g. for
+actual person accounts testing).
+
+### Simulating Networking Errors with VCR
+
+It is possible to simulate networking errors by hand-editing VCR files.
```
Comment (davidmreed) [inline docs/testing-principles.md:132]: I am not persuaded that this strategy is superior to mocking - actually I think there's a case for the reverse. Applying this strategy results in VCR artifacts being a mix of automatically-generated and handwritten, muddying the story around how they're updated. And if the hand-edited cassettes need to be updated, the hand-editing process has to be repeated.

Is there a clear way to avoid that issue? Otherwise, I'm inclined to say that we still need mocks in some cases.

File: docs/testing-principles.md (hunk 35 above)
Comment (davidmreed) [inline docs/testing-principles.md:155]:
```suggestion
client libraries by default. The case of Github3 is arguable, but simple_salesforce
```

File: docs/testing-principles.md [hunk 47]
```diff
@@ -0,0 +1,200 @@
...
+This pattern can be observed [here](https://github.com/SFDO-Tooling/CumulusCI/blob/35b07ebfa80170f6e0e9d8cabd3da75bf41d5d59/cumulusci/utils/salesforce/tests/test_count_sobjects.py#L23) and [here](https://github.com/SFDO-Tooling/CumulusCI/blob/e3ba7e8b542755b6395dbc895daee3d1ea3d11dc/cumulusci/utils/salesforce/tests/cassettes/ManualEdit_TestCountSObjects.test_count_sobjects__network_errors.yaml#L16).
+
+### Service Connectors
```
Comment (davidmreed) [inline docs/testing-principles.md:135]: I love this section, but I think it's more future-looking than a currently applicable strategy. The fact is (sadly) that we don't wrap `simple_salesforce` or `github3` today, although we manifestly should. 

I think that a prospective contributor may be confused by this section, but it has great value for us in shaping future development.

File: docs/testing-principles.md [hunk 48]
```diff
@@ -0,0 +1,200 @@
...
+sf_org.get_user_licenses()
+```
+
+The latter can be faked much more easily and scalably. Imagine if you are testing
+a large component which calls `sf_org.query()` six (or six hundred!) times. You'll
+need to use string matching (which is intrinsically fragile) to figure out which
+return values to return for which call.
+
+Between the benefits of simpler tests, cleaner code and easier replacement of abstractions,
+ServiceConnectors will pay for themselves quickly.
```
Comment (davidmreed) [inline docs/testing-principles.md:183]: These two sections seem like a very persuasive ADR in favor of using the Service Connector pattern.

**Review by davidmreed — COMMENTED** [review]
I like this doc a lot, but I'm not convinced it belongs in our public doc set. It seems to me that this should be the north star for our testing program, but only portions of it are appropriate for introducing any external contributors or new team members to _current_ best practice.

What do you think about moving all or part of this document into an ADR and/or a design document for applying the Service Connector pattern across the codebase?

File: docs/testing-principles.md (hunk 20 above)
Comment (Br4nd0R) [inline docs/testing-principles.md:76]:
```suggestion
real systems, just as we have high coverage when they are turned on.
```

File: docs/testing-principles.md [hunk 49]
```diff
@@ -0,0 +1,200 @@
...
+code for handling errors, service outages, latency and high scale.
+
+All other code should run against _real_ services or faked services
+interchangably, depending on the current step in the SDLC.
+
+### Component Tests
+
+One of the goals is to test our system as real users use it, which is why
+we use high-level APIs like `task()` instead of low level ones like
+`task._run_task()`
```
Comment (Br4nd0R) [inline docs/testing-principles.md:88]: 💯

File: docs/testing-principles.md (hunk 47 above)
Comment (davidmreed) [reply docs/testing-principles.md:137]: n.b. I've written part of a wrapper abstracting over github3 and gitlab as an experiment that we could explore furthering.

File: docs/testing-principles.md [hunk 50]
```diff
@@ -0,0 +1,202 @@
...
+
+* We make heavy use of `simple_salesforce.restful()`, which tunnels MANY different types
+   of requests through a single endpoint and is therefore hard to fake.
+
+* We make heavy use of `simple_salesforce.query()`, which has the same problem.
+
+`Github3` is better, but arguably if we had started using it by wrapping it we'd be in a
+better position to add other VCS regardless, so maybe we should just always wrap
+client libraries by default. The case of `Github3` is arguably, but `simple_salesforce`
+is not: the fact that we do all sorts of protocol-level things (passing URLs,
```
Comment (jstvz) [inline docs/testing-principles.md:158]:
```suggestion
client libraries by default. The case of `Github3` is debatable, but `simple_salesforce`
is not: the fact that we do all sorts of protocol-level things (passing URLs,
```

**Review by jstvz — COMMENTED** [review]
I think we have two documents here:

1. A document detailing our unit tests for contributors (examples from [Django](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/unit-tests/) and [CPython](https://devguide.python.org/testing/)). See, for example, the VCR section. This is a good place to include information about our custom fixtures. For example, users may not expect that they will need [workarounds](https://github.com/SFDO-Tooling/CumulusCI/blob/61ab5ad18479efd811f15522be37441130421379/cumulusci/oauth/tests/test_client.py#L269-L272) for `--replace-vcrs`. We should expand the section and include it in the public docs.
2. A document describing how test complexity indicates the wrong abstraction. There is also an argument that we should wrap third party code using a gateway. I support both ideas, but they're not a good fit for contributor-facing documentation. But there is a solid ADR here.
